source/server/session_repository.cpp

```diff
@@ -8,6 +8,16 @@ SessionRepository::SessionRepository()
 {
 }
 
+SessionRepository::~SessionRepository()
+{
+  // Do not cleanup external resources during shutdown. At this stage, all of the
+  // driver Library objects have been cleaned up so it's not safe to call them.
+  // To fix that, we'd have to change up our resource management. But right now we
+  // don't have a structured cleanup process and will let the driver figure out how
+  // to handle ctrl+c shutdown.
+  reset_server(/* cleanup_external_resources =*/false);
```

Comment:
This fixes the second mess. Trying to clean up the data structures in the destructor with default ordering was not working well and my tests exposed that (introduced by earlier cross-driver sessions change). `reset_server` can handle this ordering, so I switched to that. That needed one additional change of suppressing the external cleanup (calling back into the driver API, because those have already been unloaded).